zyngine/zynthian_engine_pianoteq_jsonrpc.py

```diff
@@ -0,0 +1,696 @@
+				options['labels'] = ['Off', 'On']
+			elif param == 'Output Mode':
+				options['labels'] = ['Stereophonic',  'Monophonic', 'Sound Recording', 'Binaural',]
+			elif param == 'Diapason':
+				pass #TODO Scale display 220..880Hz
+			if init:
+				zctrl = zynthian_controller(self, param, params[param]['name'], options)
+				self._ctrls[param] = zctrl
+				# Default MIDI CC mapping
+				default_cc = {'Sustain Pedal': 64, 'Sostenuto Pedal': 66, 'Soft Pedal': 67, 'Harmonic Pedal': 69}
+				if param in default_cc:
+					zctrl.set_midi_learn(layer.midi_chan, default_cc[param])
+			else:
+				self._ctrls[param].set_options(options)
+		return self._ctrls
+
+
+	def send_controller_value(self, zctrl):
+		self.set_param(zctrl.symbol, zctrl.value)
+
+
+	# ---------------------------------------------------------------------------
+	# API methods
+	# ---------------------------------------------------------------------------
+
+	@classmethod
+	def zynapi_get_banks(cls):
+		banks = []
+		for d in os.listdir(cls.user_presets_dpath):
+			if os.listdir(f'{cls.user_presets_dpath}/{d}'):
+				banks.append({
+					'text': d,
+					'name': d,
+					'fullpath': f'{cls.user_presets_dpath}/{d}',
+					'readonly': False
+				})
+		banks.append({
+			'text': 'Factory Presets',
+			'name': '',
+			'fullpath': '',
+			'readonly': True
+		})
+		return banks
+
+
+	@classmethod
+	def zynapi_get_presets(cls, bank):
+		presets = []
+		if bank['name'] == '':
+			all_presets = check_output([PIANOTEQ_BINARY, '--list-presets']).decode('utf-8').split('\n')
+			for preset in all_presets:
+				if preset == '' or  '/' in preset: continue
+				presets.append({
+					'text': preset,
+					'name': preset,
+					'fullpath': '',
+					'readonly': True
+				})
+		else:
+			for f in os.listdir(f"{bank['fullpath']}"):
+				if f.endswith('.fxp'):
+					presets.append({
+						'text': f,
+						'name': f[:-4],
+						'fullpath': f"{bank['fullpath']}/{f}",
+						'raw': f"{bank['fullpath']}/{f}",
+						'readonly': False
+					})
+		return presets
+
+
+	@classmethod
+	def zynapi_download(cls, fullpath):
+		return fullpath
+
+
+	@classmethod
+	def zynapi_get_formats(cls):
+		return "fxp"
+
+
+	@classmethod
+	def zynapi_install(cls, dpath, bank_path):
+		fname, ext = os.path.splitext(dpath)
+		if ext.lower() in ['.fxp']:
+			shutil.move(dpath, bank_path)
+		else:
+			raise Exception("File doesn't look like a Pianoteq FXP preset")
+
+
+	@classmethod
+	def zynapi_rename_preset(cls, preset_path, new_preset_name):
+		if preset_path[-4:].lower() != ".fxp":
+			return False
+		try:
+			head, tail = os.path.split(preset_path)
+			fname, ext = os.path.splitext(tail)
+			new_preset_path = head + "/" + new_preset_name + ext
+			os.rename(preset_path, new_preset_path)
+			return True
+		except:
+			pass
+		return False
+		
+
+	@classmethod
+	def zynapi_remove_preset(cls, preset_path):
+		if preset_path[-4:].lower() != ".fxp":
+			return False
+		os.system(f"rm '{preset_path}'")
+		return True
+
```

Comment:
Perhaps having a "zynthian_engine_pianoteq_base" class have sense, specially if we plan to keep compatibility with pianoteq 6 in the future. I know it's an extra effort, but code will be easier to maintain as we avoid duplication. Also, it's a good exercise and it normally helps to improve code quality. Pianoteq is a key engine for the project so the effort is worth it.

Reply:
We could use the new pianoteq class and derive a class from it that implements legacy (<7.5) control.

Reply:
See comment above in other conversation. There remain 2 non-inherited classes.